execution_engine/src/core/tracking_copy/mod.rs

```diff
@@ -231,6 +266,29 @@ impl<R: StateReader<Key, StoredValue>> TrackingCopy<R> {
         }
     }
 
+    pub fn get_keys(
+        &mut self,
+        correlation_id: CorrelationId,
+        key_tag: &KeyTag,
+    ) -> Result<BTreeSet<Key>, R::Error> {
+        let mut ret: BTreeSet<Key> = BTreeSet::new();
+        match self.cache.get_key_tag_reads_cached(&key_tag) {
+            Some(keys) => ret.extend(keys),
+            None => {
+                let key_tag = key_tag.to_owned();
+                let keys = self
+                    .reader
+                    .keys_with_prefix(correlation_id, &[key_tag as u8])?;
+                ret.extend(keys);
+                self.cache.insert_key_tag_read(key_tag, ret.to_owned())
+            }
+        }
+        if let Some(keys) = self.cache.get_key_tag_muts_cached(&key_tag) {
```

Comment:
I am a bit confused here should `key_tag_reads_cached` and `key_tag_muts_cached` be disjoint?  Or is the idea that `key_tag_muts_cached` masks `key_tag_reads_cached` and they can have duplicates?

Reply:
I see now; that masking happens elsewhere.  This is for key enumeration; if you mutated a key these sets will contain duplicates.

Reply:
yah